applications/StructuralMechanicsApplication/custom_response_functions/response_utilities/adjoint_nodal_displacement_response_function.cpp

```diff
@@ -64,21 +67,25 @@ namespace Kratos
             rResponseGradient.resize(rResidualGradient.size1(), false);
 
         rResponseGradient.clear();
-
-        if( rAdjointElement.Id() == mpNeighboringElement->Id() )
-        {
-            DofsVectorType dofs_of_element;
-            mpNeighboringElement->GetDofList(dofs_of_element, rProcessInfo);
-
-            const DoubleVariableType& r_traced_adjoint_dof =
-                KratosComponents<DoubleVariableType>::Get(std::string("ADJOINT_") + mTracedDofLabel);
-
-            for(IndexType i = 0; i < dofs_of_element.size(); ++i)
-            {
-                if (dofs_of_element[i]->Id() == mpTracedNode->Id() &&
-                    dofs_of_element[i]->GetVariable() == r_traced_adjoint_dof)
-                {
-                    rResponseGradient[i] = -1;
+        const Variable<double>* adjoint_solution_variable = &KratosComponents<Variable<double>>::Get("ADJOINT_" + mTracedDofLabel + "_X");
+
+        std::map<IndexType, IndexType>::iterator it;
+        for (it = mNodeElementMap.begin(); it != mNodeElementMap.end(); ++it) {
+            const auto node_id = it->first;
+            const auto element_id = it->second;
+
+            if( rAdjointElement.Id() == element_id ) {
+                DofsVectorType dofs_of_element;
```

Comment:
I would move this out of the loop to reuse it

Reply:
but then it is created for all elements of the system. Is this really better? Especially in cases with a much higher number of elements of the system as traced neighboring elements.

Reply:
If the vector is created outside the loop it is only created once. Then inside the loop it does not have to be allocated all the time, because it might already have the correct size. Its the same logic as with the assembly process and why the local LHS matrix is passed to the element, but not created inside the elements CalculateLHS function itself.